src/writer/update-proposal.ts

```diff
@@ -0,0 +1,147 @@
+
+export function isScamDetected({ address, name, body }): boolean {
+  const proposalNameLC = name.toLowerCase();
+  const proposalBodyLC = body.toLowerCase();
+  const addressLC = address.toLowerCase();
+
+  return (
+    flaggedAddresses.includes(addressLC) ||
+    flaggedProposalBodyKeywords.some(keyword => proposalBodyLC.includes(keyword)) ||
+    flaggedProposalTitleKeywords.some(keyword => proposalNameLC.includes(keyword))
+  );
+}
+
+export function isAddressAuthorized(address: string, space: any): boolean {
+  const addressLC = address.toLowerCase();
+  const members = [
+    ...(space.members || []),
+    ...(space.admins || []),
+    ...(space.moderators || [])
+  ].map(member => member.toLowerCase());
+
+  return members.includes(addressLC);
+}
+
+export async function verify(body): Promise<any> {
+  const msg = jsonParse(body.msg);
+  const timestampNow = Math.floor(Date.now() / 1e3);
+
+  const schemaIsValid = snapshot.utils.validateSchema(snapshot.schemas.updateProposal, msg.payload);
+  if (!schemaIsValid) {
+    log.warn('[writer] Wrong proposal format', schemaIsValid);
+    return Promise.reject('wrong proposal format');
+  }
+
+  const proposal = await getProposal(msg.space, msg.payload.proposal);
+  if (!proposal) return Promise.reject('unknown proposal');
+  if (proposal.start < timestampNow) return Promise.reject('proposal already started');
+
+  const isChoicesValid = validateChoices({
+    type: msg.payload.type || proposal.type,
+    choices: msg.payload.choices.length > 0 ? msg.payload.choices : proposal.choices
+  });
+  if (!isChoicesValid) {
+    return Promise.reject(`wrong choices for "${msg.payload.type || proposal.type}" type voting`);
+  }
+
+  if (proposal.author !== body.address) return Promise.reject('Not the author');
+
+  const space = await getSpace(msg.space);
+  space.id = msg.space;
+
+  const spaceUpdateError = getSpaceUpdateError({
+    type: msg.payload.type || proposal.type,
+    space
+  });
+  if (spaceUpdateError) return Promise.reject(spaceUpdateError);
+
+  const hasScam = isScamDetected({
+    address: body.address,
+    name: msg.payload.name || proposal.title,
+    body: msg.payload.body || proposal.body
+  });
+  if (hasScam) return Promise.reject('scam proposal detected, contact support');
+
+  return Promise.resolve(proposal);
+}
+
+export async function action(body, ipfs, receipt, id, context): Promise<void> {
+  const originalProposal = context;
+  const msg = jsonParse(body.msg);
+  const updated = parseInt(msg.timestamp);
+  const metadata = msg.payload.metadata || {};
+  const plugins = JSON.stringify(metadata.plugins || {});
+
+  const proposal = {
+    ipfs,
+    updated,
+    type: msg.payload.type || originalProposal.type,
+    plugins: plugins || originalProposal.plugins,
+    title: msg.payload.name || originalProposal.title,
+    body: msg.payload.body,
+    discussion: msg.payload.discussion,
+    choices: JSON.stringify(msg.payload.choices) || originalProposal.choices
+  };
```

Comment:
What does the UI looks like ? Isn't it coming from the same form as the create proposal form, in which case we should expect the full proposal, and don't really need to fallback on `originalProposal` ?

Reply:
UI is not a part of this PR, you can check all details [here](https://www.notion.so/snapshotlabs/Proposal-update-backend-c5f55c83f1ff4f3b82a273e96cb9c463?pvs=4) and the task for UI is [here](https://www.notion.so/snapshotlabs/Allow-proposal-update-during-delay-period-289e6209cff5401f90fdb9c865efaf81?pvs=4)

Reply:
But what kind of payload are we expecting from UI ? The same as create proposal ? or if the user only edit the title, we only received the title ?

Reply:
no, it will be a different payload. You can check it [here](https://github.com/snapshot-labs/snapshot.js/pull/895/files#diff-0de6060ecf3e628ed7a2ef320e36093f395de48084a4c2a4c2977588b6c0011dR143) and [here](https://github.com/snapshot-labs/snapshot.js/pull/895/files#diff-ff4da39e64c3be82c61a689b48eff52337efa570ec09f45be4f0f945f6f9097cR1)